### PR TITLE
[PROD-18994] Initial client server code

### PR DESCRIPTION
**Pull request by patricia-bluecore**
Build rate limiter that accepts GRPC / PRPC requests. The limiter leverages Go's rate/limit package to track the next available time to schedule emails. It stores this information in an in-memory data structure. This is a more efficient behavior than the server waiting until the customer is under their rate limit.

### CODE REVIEW

File: server/server.go [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+	visitors[namespace].next_available_time = reservation_time
+
+	return reservation_time
+}
+
+
+func (s *server) GetFutureScheduleTime(ctx context.Context, request *proto.GetFutureScheduleTimeRequest) (*proto.GetFutureScheduleTimeResponse, error) {
+
+	limit := int(request.GetRateLimitPerSecond())
+	email_count := int(request.GetEmailCount())
```
Comment (kshahat) [inline server/server.go]: `emailCount`

File: server/server.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	namespace := request.GetNamespace()
```
Comment (kshahat) [inline server/server.go]: since you're passing `Esp` you should be using that as well. I'd create a function `createKey` that takes the namespace and esp, lowercases both of them, and combines them in some way to create a key and just use `key` from this point on.

File: server/server.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+	mtx.Lock()
+	reservation_time := makeReservation(namespace, limit, email_count)
+	mtx.Unlock()
+
+	reserved, err := ptypes.TimestampProto(reservation_time)
+
+	if err != nil {
+		panic(err)
```
Comment (kshahat) [inline server/server.go]: You shouldn't be panicking here. You should log the error that happened and `return nil, err` and let main figure out what to do in this case

File: server/server.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+
+func (s *server) GetFutureScheduleTime(ctx context.Context, request *proto.GetFutureScheduleTimeRequest) (*proto.GetFutureScheduleTimeResponse, error) {
+
+	limit := int(request.GetRateLimitPerSecond())
+	email_count := int(request.GetEmailCount())
+	namespace := request.GetNamespace()
+
+	mtx.Lock()
```
Comment (kshahat) [inline server/server.go]: I would move the lock/unlock combo within the function, make this function a method, and at the very beginning do 
```
mtx.Lock()
defer mtx.Unlock()
```
This has the benefit of only using the mtx in the function that updates the map and will Unlock the mutex on leaving the scope in all cases (early returns and otherwise). But I think maybe this use of mtx might limit our throughput too much. Let's discuss this

File: server/server.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	fmt.Println("connecting to http://localhost:8080")
+
+	listener, err := net.Listen("tcp", ":8080")
+	if err != nil {
+		panic(err)
+	}
+
+	srv := grpc.NewServer()
+	proto.RegisterRateLimitServer(srv, &server{})
+	reflection.Register(srv)
```
Comment (kshahat) [inline server/server.go]: This is fine for now/testing but eventually this should create an http port for prpc

File: server/server.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	resp := &proto.GetFutureScheduleTimeResponse{ScheduledSendTime: reserved}
+
+	return resp, nil
+}
+
+func main() {
```
Comment (kshahat) [inline server/server.go]: this function should be in it's own package in the root directory as package `main`

File: service/_unit_test.go [hunk 7]
```diff
@@ -0,0 +1,20 @@
+package main
```
Comment (kshahat) [inline service/_unit_test.go]: Since this is in the `service` directory with the `service` package, this should also be `package service`. The file name should then become `service_test.go` since we assume that all non-integration tests are unit tests

File: service/_unit_test.go [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+	"reflect"
+	"testing"
+)
+
+func TestAddLimiter(t *testing.T) {
+	key := "express.cheetah_mail"
+	limit := 5
+	actual := AddLimiter(key, limit)
+
+	now := actual.next_available_time
```
Comment (kshahat) [inline service/_unit_test.go]: I'm not sure the times will be exactly in sync when you do the comparison because there may be a time difference?

File: service/service.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+	limiter           *rate.Limiter
+	nextAvailableTime time.Time
+}
+
+func New() *Server {
+	return &Server{mtx: sync.Mutex{}, clients: make(map[string]*clientLimit)}
+}
+
+func createKey(namespace string, esp string) string {
+	return strings.ToLower(namespace) + "." + strings.ToLower(esp)
```
Comment (kshahat) [inline service/service.go]: 💯

File: service/service.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+	clients map[string]*clientLimit
+}
+
+type clientLimit struct {
+	limiter           *rate.Limiter
+	nextAvailableTime time.Time
+}
+
+func New() *Server {
+	return &Server{mtx: sync.Mutex{}, clients: make(map[string]*clientLimit)}
```
Comment (kshahat) [inline service/service.go]: 💯

File: service/service.go [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+	mtx     sync.Mutex
+	clients map[string]*clientLimit
+}
+
+type clientLimit struct {
+	limiter           *rate.Limiter
+	nextAvailableTime time.Time
+}
+
+func New() *Server {
```
Comment (kshahat) [inline service/service.go]: For files that are externally used, you should have a description above the function definition that begins with the function name. In this example it would be `//New blah blah blah and will return blah blah which means blah`

File: service/service.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+	now := time.Now()
+
+	clientLimitInstance := &clientLimit{limiter: limiter, nextAvailableTime: now}
+
+	s.clients[key] = clientLimitInstance
+	return clientLimitInstance
+
+}
+
+func (s *Server) Lock() {
```
Comment (kshahat) [inline service/service.go]: Let's discuss this in person. The current mutex paradigm may need to shift a bit

File: service/service.go [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+type clientLimit struct {
+	limiter           *rate.Limiter
+	nextAvailableTime time.Time
+}
+
+func New() *Server {
+	return &Server{mtx: sync.Mutex{}, clients: make(map[string]*clientLimit)}
+}
+
+func createKey(namespace string, esp string) string {
```
Comment (kshahat) [inline service/service.go]: There's a cool thing you can do in go: 
`func foo(s1, s2, s3 string, int1, int2, int3 int, b1, b2 bool)`
which specifies that 
- s1, s2, & s3 are of type string
- int1, int2, & int3 are of type int 
- b1 & b2 are of type bool

your way 100% works but you can condense it

File: service/service.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+	return clientLimitInstance
+
+}
+
+func (s *Server) Lock() {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+}
+
+func makeReservation(s *Server, key string, limit int, emailCount int) time.Time {
```
Comment (kshahat) [inline service/service.go]: Why not make this a member function if it operates on the Server itself? This function should also probably return an error as well when reservation.OK() is false that we can propagate back to the client to have them reschedule batches in the future.

This function should have more comments to explain to someone reading the code what is happening as well.

File: service/service.go [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+	delay := reservation.DelayFrom(client.nextAvailableTime)
+
+	reservationTime := client.nextAvailableTime.Add(delay)
+
+	s.clients[key].nextAvailableTime = reservationTime
+
+	return reservationTime
+}
+
+func (s *Server) GetFutureScheduleTime(ctx context.Context, request *ratelimit.GetFutureScheduleTimeRequest) (*ratelimit.GetFutureScheduleTimeResponse, error) {
```
Comment (kshahat) [inline service/service.go]: Add function comment for this function as well, please!

File: service/service.go [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+	clientLimitInstance := &clientLimit{limiter: limiter, nextAvailableTime: now, mtx: sync.Mutex{}}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	s.clients[key] = clientLimitInstance
+
+	return clientLimitInstance
+}
+
+func (s *Server) makeReservation(key string, limit int, emailCount int) (time.Time, error) {
```
Comment (kshahat) [inline service/service.go]: should this function return an error if `emailCount > limit`?

File: service/service_test.go [hunk 17]
```diff
@@ -0,0 +1,59 @@
+package service
+
+import (
+	"golang.org/x/time/rate"
+	"testing"
+	"time"
+)
+
+func TestMakeReservation(test *testing.T) {
```
Comment (kshahat) [inline service/service_test.go]: This is a really good test. Good job!

File: service/service.go [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+	// be the bucket size where r=b. The limiter makes the assumption that client applications will chunk
+	// emails so that emailCount is less than or equal to the rate limit.
+	burst := limit
+	rateLimit := rate.Limit(limit)
+	limiter := rate.NewLimiter(rateLimit, burst)
+
+	now := time.Now()
+	clientLimitInstance := &clientLimit{limiter: limiter, nextAvailableTime: now, mtx: sync.Mutex{}}
+
+	s.mtx.Lock()
```
[the remark this answers is not in the document]
Comment (patricia-bluecore) [reply service/service.go]: the rational was you only need to lock it before you write to the map but i can put it on the top if it's more clearer!

[the remark this answers is not in the document]
Comment (kshahat) [reply service/service.go]: Yeah I think it's best to do that because the rest of the function is negligible in terms of work to be done

File: main.go [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+
+	listener, err := net.Listen("tcp", ":8080")
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	srv := grpc.NewServer()
+	service := service.New()
+	ratelimit.RegisterRateLimitServer(srv, service)
+	reflection.Register(srv)
```
Comment (kshahat) [inline main.go]: I think this will work fine without this `reflection.Register` thing because you're basically doing that the line before. Also this will not be communicating over grpc but rather prpc so you'll have to change that. You can make this change in a separate PR.

File: service/service.go [hunk 20]
```diff
@@ -0,0 +1,132 @@
...
+	}
+
+	client.mtx.Lock()
+	defer client.mtx.Unlock()
+
+	if client.limiter.Burst() != limit {
+		client = s.newLimiterForKey(key, limit)
+	}
+
+	if emailCount > limit {
```
Comment (kshahat) [inline service/service.go]: Can you move this to be the first thing in the function? It would make sense since it's an invalid argument.

Actually, now that I think about it, this should probably be in `GetFutureScheduleTime`. Normally a server would confirm that it's parameters are correct/fixing them if needed at the beginning and allowing the rest of the code to make the assumptions it needs to later on in the process. In the case of this application, ensuring that `emailCount <= limit` and that the string parameters are normalized (lowercased). Apologies if this leads to any confusions!
